File: examples/basic_prediction.py
"""Minimal prediction, planning, and evaluation example."""

import json

from worldforge import Action, BBox, Position, SceneObject, WorldForge


def main() -> None:
    forge = WorldForge()
    world = forge.create_world("kitchen", provider="mock")

    world.add_object(
        SceneObject(
            "red_mug",
            Position(0.0, 0.8, 0.0),
            BBox(Position(-0.05, 0.75, -0.05), Position(0.05, 0.85, 0.05)),
        )
    )

    prediction = world.predict(Action.move_to(0.3, 0.8, 0.0), steps=2)
    plan = world.plan(goal="move the mug to the right")

    print(
        json.dumps(
            {
                "prediction": {
                    "provider": prediction.provider,
                    "physics_score": prediction.physics_score,
                    "confidence": prediction.confidence,
                },
                "plan": {
                    "provider": plan.provider,
                    "actions": plan.action_count,
                    "success_probability": plan.success_probability,
                },
            },
            indent=2,
            sort_keys=True,
        )
    )

    report = world.evaluate("physics")
    print(report.to_markdown())


if __name__ == "__main__":
    main()

File: examples/basic_prediction.rs
//! Basic prediction example using WorldForge.
//!
//! This example demonstrates how to create a world and run a single prediction
//! using the auto-detected provider registry.
//!
//! # Running
//!
//! ```bash
//! cargo run --example basic_prediction
//! ```

use worldforge_core::{Action, PredictionConfig};
use worldforge_providers::auto_detect_worldforge;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Auto-detect all available providers from environment variables.
    // The Mock provider is always available.
    let wf = auto_detect_worldforge();

    // List available providers
    let providers = wf.list_providers();
    println!("Available providers: {:?}", providers);

    // Create a world using the mock provider (always available)
    let world = wf.create_world("kitchen", "mock")?;
    println!("Created world: kitchen (provider: mock)");

    // Define an action: move an object to coordinates (0.5, 0.8, 0.0)
    let action = Action::move_to(0.5, 0.8, 0.0);

    // Run a prediction for 10 steps
    let config = PredictionConfig {
        steps: 10,
        ..Default::default()
    };
    let prediction = world.predict(&action, &config).await?;

    // Print results
    println!("Prediction complete:");
    println!("  Physics score: {:.2}", prediction.physics_score);
    println!("  Frames generated: {}", prediction.frames.len());

    Ok(())
}
